src/workerd/api/streams/writable-sink-adapter.c++

```diff
@@ -0,0 +1,864 @@
+    while (!queue.empty() && !aborted) {
+      auto task = KJ_ASSERT_NONNULL(queue.pop());
+      KJ_DEFER({
+        if (task->fulfiller->isWaiting()) {
+          KJ_IF_SOME(pending, pendingAbort) {
+            task->fulfiller->reject(kj::mv(pending));
+          } else {
+            task->fulfiller->reject(KJ_EXCEPTION(DISCONNECTED, "Task was canceled."));
+          }
+        }
+      });
+      bool taskFailed = false;
+      try {
+        co_await task->task();
+        task->fulfiller->fulfill();
+      } catch (...) {
+        auto ex = kj::getCaughtExceptionAsKj();
+        task->fulfiller->reject(kj::mv(ex));
+        taskFailed = true;
+      }
+      // If the task failed, we exit the loop. We're going to abort the
+      // entire remaining queue anyway so there's no point in continuing.
+      if (taskFailed) co_return;
+    }
+  }
+};
+
+WritableStreamSinkJsAdapter::WritableStreamSinkJsAdapter(jsg::Lock& js,
+    IoContext& ioContext,
+    kj::Own<WritableStreamSink> sink,
+    kj::Maybe<Options> options)
+    : state(ioContext.addObject(kj::heap<Active>(kj::mv(sink), kj::mv(options).orDefault({})))),
+      backpressureState(newBackpressureState(js)),
+      selfRef(kj::rc<WeakRef<WritableStreamSinkJsAdapter>>(
+          kj::Badge<WritableStreamSinkJsAdapter>{}, *this)) {
+  // We want the initial backpressure state to be "ready".
+  backpressureState.release(js);
+}
+
+WritableStreamSinkJsAdapter::WritableStreamSinkJsAdapter(jsg::Lock& js,
+    IoContext& ioContext,
+    kj::Own<kj::AsyncOutputStream> stream,
+    StreamEncoding encoding,
+    kj::Maybe<Options> options)
+    : WritableStreamSinkJsAdapter(
+          js, ioContext, newSystemStream(kj::mv(stream), encoding, ioContext), kj::mv(options)) {}
+
+WritableStreamSinkJsAdapter::~WritableStreamSinkJsAdapter() noexcept(false) {
+  selfRef->invalidate();
+}
+
+kj::Maybe<const kj::Exception&> WritableStreamSinkJsAdapter::isErrored() {
+  return state.tryGet<kj::Exception>();
+}
+
+bool WritableStreamSinkJsAdapter::isClosed() {
+  return state.is<Closed>();
+}
+
+bool WritableStreamSinkJsAdapter::isClosing() {
+  KJ_IF_SOME(active, state.tryGet<IoOwn<Active>>()) {
+    return active->closePending;
+  } else {
+    return false;
+  }
+}
+
+kj::Maybe<ssize_t> WritableStreamSinkJsAdapter::getDesiredSize() {
+  return state.tryGet<IoOwn<Active>>().map(
+      [](IoOwn<Active>& active) { return active->getDesiredSize(); });
+}
+
+jsg::Promise<void> WritableStreamSinkJsAdapter::write(jsg::Lock& js, const jsg::JsValue& value) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(activeOwn, IoOwn<Active>) {
+      // Deference the IoOwn once to get the active state.
+      auto& active = *activeOwn;
+
+      // If close is pending, we cannot accept any more writes.
+      if (active.closePending) {
+        auto exc = js.typeError("Write after close is not allowed");
+        // If we are configured to treat early errors as fatal, we need to
+        // abort the entire adapter.
```

Comment:
This comment describes an action that doesn't seem to exist. Should we `aborted = true` or something?

Reply:
Left over comment from an early iteration.